test/Router.t.sol

```diff
@@ -156,6 +161,63 @@ contract RouterTest is Test, LogicSignature, DelegationSignature {
         if (agent != executingAgent) revert();
     }
 
+    function testExecuteBySig() external {
+        vm.prank(user);
+        router.newAgent();
+        IParam.Logic[] memory logics = new IParam.Logic[](1);
+        logics[0] = IParam.Logic(
+            address(mockTo), // to
+            '',
+            inputsEmpty,
+            IParam.WrapMode.NONE,
+            address(0), // approveTo
+            address(0) // callback
+        );
+        // Ensure correct EIP-712 encodeData
+        uint256 deadline = block.timestamp + 3600;
+        uint256 nonce = 0;
+        IParam.ExecutionDetails memory details = IParam.ExecutionDetails(
+            logics,
+            tokensReturnEmpty,
+            SIGNER_REFERRAL,
+            nonce,
+            deadline
+        );
+        bytes memory signature = getTypedDataSignature(details, router.domainSeparator(), userPrivateKey);
+        vm.expectEmit(true, true, true, true, address(router));
+        emit Execute(user, address(router.agents(user)), SIGNER_REFERRAL);
+        router.executeBySig(details, user, signature);
+        uint256 newNonce = router.executionNonces(user);
+        assertEq(newNonce, nonce + 1);
+    }
+
+    function testCannotExecuteBySigWithIncorrectSignature() external {
+        vm.prank(user);
+        router.newAgent();
+        IParam.Logic[] memory logics = new IParam.Logic[](1);
+        logics[0] = IParam.Logic(
+            address(mockTo), // to
+            '',
+            inputsEmpty,
+            IParam.WrapMode.NONE,
+            address(0), // approveTo
+            address(0) // callback
+        );
+        // Ensure correct EIP-712 encodeData
+        uint256 deadline = block.timestamp + 3600;
+        uint256 nonce = 0;
+        IParam.ExecutionDetails memory details = IParam.ExecutionDetails(
+            logics,
+            tokensReturnEmpty,
+            SIGNER_REFERRAL,
+            nonce,
+            deadline
+        );
+        bytes memory signature = getTypedDataSignature(details, router.domainSeparator(), signerPrivateKey);
+        vm.expectRevert(IRouter.InvalidSignature.selector);
+        router.executeBySig(details, user, signature);
```

Comment:
This test is aimed at evaluating different private keys rather than the content of `IParam.ExecutionDetails`, unlike `testCannotExecuteInvalidSignature`. Would it make sense to only retain the following 3 lines for this test?

```suggestion
        bytes memory signature = getTypedDataSignature(executionDetailsEmpty, router.domainSeparator(), signerPrivateKey);
        vm.expectRevert(IRouter.InvalidSignature.selector);
        router.executeBySig(executionDetailsEmpty, user, signature);
```

Reply:
I think so. Good point.

Reply:
Have a different thought when modifying the code. This test case is exactly the same as `testExecuteBySig` above, with only the signer different right now. If we want to change those to empty ones, will it be more proper to change all the test cases at the same time?

Reply:
I'm thinking should this test tamper values in `ExecutionDetails` like `testCannotExecuteInvalidSignature`, to confirm the implementation of `TypedDataHash.sol` is corrent? Using `signerPrivateKey` to sign seems like we're just checking if OpenZeppelin's `isValidSignatureNow` works.

Reply:
The correctness of hash value is tested in `ExecutionTypeHash.t.sol`. The test cases here is more about if the logic and signer checks are correct.

Reply:
I think you make a valid point. The tests for EIP-712 `typehash` & `_hash` should be centralized in the `*Typehash.t.sol` file. It seems redundant for Router.t.sol to have these many checks. It looks like we can review all tests with the `// Ensure correct EIP-712 encodeData` comment and replace with `*Empty` for clearer understanding.

Also, could you please upload the codes that generates sig for `DelegationTypehash.t.sol` & `ExecutionTypehash.t.sol`? Something like this:
https://github.com/dinngo/test-dapp/tree/for-protocolink-contract-delegation.

Reply:
> It looks like we can review all tests with the // Ensure correct EIP-712 encodeData comment and replace with *Empty for clearer understanding.

Will modify the test cases then.

> Also, could you please upload the codes that generates sig for DelegationTypehash.t.sol & ExecutionTypehash.t.sol?

I don't think I kept all the codes but will do that later.